crates/nu-protocol/src/value/record.rs

```diff
@@ -36,10 +48,68 @@ impl Record {
         usize::min(self.cols.len(), self.vals.len())
     }
 
+    /// Naive push to the end of the datastructure.
+    ///
+    /// May duplicate data!
+    ///
+    /// Consider to use [`Record::insert`] instead
     pub fn push(&mut self, col: impl Into<String>, val: Value) {
         self.cols.push(col.into());
         self.vals.push(val);
     }
+
+    /// Insert into the record, replacing preexisting value if found.
+    ///
+    /// Returns `Some(previous_value)` if found. Else `None`
+    pub fn insert(&mut self, col: Cow<str>, val: Value) -> Option<Value> {
+        if let Some(idx) = self.columns().position(|k| k == &col) {
+            // Can panic if vals.len() < cols.len()
+            let curr_val = &mut self.vals[idx];
+            Some(std::mem::replace(curr_val, val))
+        } else {
+            self.cols.push(col.into_owned());
+            self.vals.push(val);
+            None
+        }
+    }
+
+    pub fn contains(&self, col: impl AsRef<str>) -> bool {
+        self.cols.iter().any(|k| k == col.as_ref())
```

Comment:
What about making a variable `let col_ref = col.as_ref()`, instead of calling `col.as_ref()` inside closure everytime?

```rust
let col_ref = col.as_ref();
self.cols.iter().any(|k| k == col_ref)
```
It can be a performance improvement if we call the function many times.

Reply:
I assumed that `.as_ref()` is pretty transparent to the optimizer but we could lift that out. Then there might be some optimizer luck necessary so it can short circuit that if `cols` is empty but we shouldn't worry too much without profiling.

I think there is an aspect that makes me dislike `AsRef` slightly, so I might look into `Borrow` as a replacement.